modules/contact/src/ContactAction.C

```diff
@@ -97,112 +108,296 @@ ContactAction::ContactAction(const InputParameters & params)
+      auto master_params = _factory.getValidParams("LowerDBlockFromSidesetGenerator");
+      auto slave_params = _factory.getValidParams("LowerDBlockFromSidesetGenerator");
+
+      master_params.set<MeshGeneratorName>("input") = _mesh_gen_name;
+      slave_params.set<MeshGeneratorName>("input") = master_name;
+
+      master_params.set<SubdomainName>("new_block_name") = master_subdomain_name;
+      slave_params.set<SubdomainName>("new_block_name") = slave_subdomain_name;
+
+      master_params.set<std::vector<BoundaryID>>("sidesets") = {_mesh->getBoundaryID(_master)};
+      slave_params.set<std::vector<BoundaryID>>("sidesets") = {_mesh->getBoundaryID(_slave)};
+
+      _app.addMeshGenerator("LowerDBlockFromSidesetGenerator", master_name, master_params);
+      _app.addMeshGenerator("LowerDBlockFromSidesetGenerator", slave_name, slave_params);
+    }
   }
 
-  // Determine number of dimensions
-  const unsigned int ndisp = displacements.size();
+  if (_current_task == "add_variable")
+  {
+    // Add the lagrange multiplier on the slave subdomain.
+    const auto addLagrangeMultiplier =
+        [this, &slave_subdomain_name, &displacements](const std::string & variable_name,
+                                                      const int codimension) //
+    {
+      InputParameters params = _factory.getValidParams("MooseVariableBase");
 
-  // convert vector of VariableName to vector of VariableName
-  std::vector<VariableName> coupled_displacements(ndisp);
-  for (unsigned int i = 0; i < ndisp; ++i)
-    coupled_displacements[i] = displacements[i];
+      mooseAssert(_problem->systemBaseNonlinear().hasVariable(displacements[0]),
+                  "Displacement variable is missing");
+      const auto primal_type =
+          _problem->systemBaseNonlinear().system().variable_type(displacements[0]);
+      const int lm_order = primal_type.order.get_order() - codimension;
 
-  if (_current_task == "add_dirac_kernel")
+      if (primal_type.family == MONOMIAL || (primal_type.family == LAGRANGE && lm_order < 1))
+      {
+        params.set<MooseEnum>("family") = "MONOMIAL";
+        params.set<MooseEnum>("order") = "CONSTANT";
+      }
+      else if (primal_type.family == LAGRANGE)
+      {
+        params.set<MooseEnum>("family") = Utility::enum_to_string<FEFamily>(primal_type.family);
+        params.set<MooseEnum>("order") = Utility::enum_to_string<Order>(OrderWrapper{lm_order});
+      }
+      else
+        mooseError("Primal variable type must be either MONOMIAL or LAGRANGE for mortar contact.");
+
+      params.set<std::vector<SubdomainName>>("block") = {slave_subdomain_name};
+      auto fe_type = AddVariableAction::feType(params);
+      auto var_type = AddVariableAction::determineType(fe_type, 1);
+      _problem->addVariable(var_type, variable_name, params);
+    };
+
+    // Set the family/order same as primal for normal contact and one order lower for tangential.
```

Comment:
Is this correct? I thought this was supposed to be at least one order lower for both normal and tangential?

Reply:
Yes, the intended behavior is same order as primal for normal and one lower for tangential. The discussion is up at the top of the thread:
https://github.com/idaholab/moose/pull/14049#issuecomment-532304258

Reply:
Gotcha, missed that comment, and I was going off of https://www.mooseframework.org/syntax/Constraints/index.html

Reply:
See also the responses [here](https://groups.google.com/d/msg/moose-users/JA1yt77MEV0/LX_CgfmwCgAJ)

Reply:
The email list was what confused me, since it sounded like you recommend one order lower for thermal, normal, and tangential. Careful re-read is to use an order lower for thermal and tangential, since they are mortar, but same order for node to face discretization.

Reply:
Yes sir! Although at some point I would really like to do some careful study of the node-face discretization with MMS and explore the finite element convergence rates.